Data/IDProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Diagnostics;
+using Anvil.CSharp.Logging;
+
+namespace Anvil.CSharp.Data
+{
+    /// <summary>
+    /// An instance that provides a unique IDs each time requested.
+    /// Includes a mechanism to detect when ID supply is near exhaustion.
+    /// </summary>
+    public class IDProvider
+    {
+        /// <summary>
+        /// Triggers the first time that any instance of <see cref="IDProvider"/> passes its
+        /// <see cref="SupplyWarningThreshold"/> for the first time.
+        /// This gives the application the opportunity to react before IDs are exhausted.
+        /// </summary>
+        public static event Action<IDProvider> OnIDLimitGlobalWarning;
+
+        /// <summary>
+        /// A reserved ID to represent "No ID".
+        /// <see cref="GetNextID"/> will never return this value even when IDs roll over and are no longer unique.
+        /// </summary>
+        public const int UNSET_ID = 0;
+
+        /// <summary>
+        /// Triggers the first time that <see cref="SupplyWarningThreshold"/> is passed.
+        /// This gives the consumer the opportunity to react before IDs are exhausted.
+        /// </summary>
+        public event Action OnIDLimitWarning;
+
+        public readonly uint SupplyWarningThreshold;
+
+        private bool m_HasIDWarningTriggered = false;
+        private uint m_LastNewID = 0;
+
+
+        /// <summary>
+        /// Creates a new ID provider optionally allowing a supply warning threshold to be set.
+        /// </summary>
+        /// <param name="supplyWarningThreshold">
+        /// (default: <see cref="uint.MaxValue"/>) The threshold which, when passed, triggers
+        /// <see cref="OnIDLimitImminent"/>.
+        /// </param>
+        public IDProvider(uint supplyWarningThreshold = uint.MaxValue-1)
```

Comment:
```suggestion
        public IDProvider(uint supplyWarningThreshold = uint.MaxValue-1_000_000)
```
On the one hand, any threshold is exceedingly unlikely to be hit - but on the other hand, why not leave a huge amount of wiggle room by default? For any arbitrary system, one million ID's remaining seems like plenty of time to refresh the system, but you still have to burn through 99.977% of ID's to get there!

Reply:
That seems reasonable to me

Reply:
Sure, 1MM is actually what I chose for the threshold in the game. I was thinking that these would generally be created in a much more rigid context that was able to abort more quickly but you make a good point. I've removed the parameter from `UpdateHandle`'s usage of `IDProvider`

Fixed